File: tests/wasm/test_decycler.js

import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import {
    loadTestData,
    assertArrayClose,
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN,
    EXPECTED_OUTPUTS
} from './test_utils.js';
import { compareWithRust } from './rust-comparison.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

test.before(async () => {

    try {
        const wasmPath = path.join(__dirname, '../../pkg/vector_ta.js');
        const importPath = process.platform === 'win32'
            ? 'file:///' + wasmPath.replace(/\\/g, '/')
            : wasmPath;
        wasm = await import(importPath);

    } catch (error) {
        console.error('Failed to load WASM module. Run "wasm-pack build --features wasm --target nodejs" first');
        throw error;
    }

    testData = loadTestData();
});

test('Decycler partial params', () => {

    const close = new Float64Array(testData.close);

    const result = wasm.decycler_js(close, 125, 0.707);
    assert.strictEqual(result.length, close.length);
});

test('Decycler accuracy', async () => {

    const close = new Float64Array(testData.close);
    const expected = EXPECTED_OUTPUTS.decycler;

    const result = wasm.decycler_js(
        close,
        expected.defaultParams.hp_period,
        expected.defaultParams.k
    );

    assert.strictEqual(result.length, close.length);


    const last5 = result.slice(-5);
    assertArrayClose(
        last5,
        expected.last5Values,
        1e-8,
        "Decycler last 5 values mismatch"
    );


    await compareWithRust('decycler', result, 'close', expected.defaultParams);
});

test('Decycler default candles', () => {

    const close = new Float64Array(testData.close);

    const result = wasm.decycler_js(close, 125, 0.707);
    assert.strictEqual(result.length, close.length);
});

test('Decycler zero period', () => {

    const inputData = new Float64Array([10.0, 20.0, 30.0]);

    assert.throws(() => {
        wasm.decycler_js(inputData, 0, 0.707);
    }, /invalid.*period/i);
});

test('Decycler period exceeds length', () => {

    const dataSmall = new Float64Array([10.0, 20.0, 30.0]);

    assert.throws(() => {
        wasm.decycler_js(dataSmall, 10, 0.707);
    }, /invalid period/i);
});

test('Decycler edge case k=0', () => {

    const inputData = new Float64Array([10.0, 20.0, 30.0, 40.0, 50.0]);

    assert.throws(() => {
        wasm.decycler_js(inputData, 3, 0.0);
    }, /invalid k/i);
});

test('Decycler NaN handling', () => {

    const close = new Float64Array(testData.close);

    const result = wasm.decycler_js(close, 125, 0.707);
    assert.strictEqual(result.length, close.length);


    const firstNonNaN = result.findIndex(v => !isNaN(v));
    if (firstNonNaN !== -1 && result.length > firstNonNaN + 127) {
        for (let i = firstNonNaN + 127; i < result.length; i++) {
            assert(!isNaN(result[i]), `Found unexpected NaN at index ${i}`);
        }
    }
});

test('Decycler warmup period', () => {

    const close = new Float64Array(testData.close);
    const hp_period = 125;

    const result = wasm.decycler_js(close, hp_period, 0.707);


    const firstNonNaN = result.findIndex(v => !isNaN(v));
    if (firstNonNaN !== -1) {


        const firstInput = close.findIndex(v => !isNaN(v));
        const expectedWarmup = firstInput + 2;

        assert.strictEqual(firstNonNaN, expectedWarmup,
            `Warmup period mismatch: expected ${expectedWarmup}, got ${firstNonNaN}`);


        for (let i = 0; i < expectedWarmup; i++) {
            assert(isNaN(result[i]), `Expected NaN at index ${i} during warmup`);
        }
    }
});

test('Decycler partial NaN input', () => {


    const data = new Float64Array(100);
    for (let i = 0; i < 100; i++) {
        data[i] = i + 1.0;
    }


    data[10] = NaN;
    data[11] = NaN;


    const result = wasm.decycler_js(data, 5, 0.707);
    assert.strictEqual(result.length, data.length);



});

test('Decycler edge case k values', () => {

    const data = new Float64Array(100);
    for (let i = 0; i < 100; i++) {
        data[i] = i + 1.0;
    }


    const resultSmall = wasm.decycler_js(data, 10, 0.001);
    assert.strictEqual(resultSmall.length, data.length);


    const resultDefault = wasm.decycler_js(data, 10, 0.707);
    assert.strictEqual(resultDefault.length, data.length);


    const resultLarge = wasm.decycler_js(data, 10, 10.0);
    assert.strictEqual(resultLarge.length, data.length);



    const warmupEnd = 2;
    const checkIdx = warmupEnd + 5;
    if (data.length > checkIdx) {
        assert(resultSmall[checkIdx] !== resultDefault[checkIdx],
            "Different k values should produce different results");
        assert(resultDefault[checkIdx] !== resultLarge[checkIdx],
            "Different k values should produce different results");
    }
});

test('Decycler reinput', () => {

    const close = new Float64Array(testData.close);
    const hp_period = 30;
    const k = 0.707;


    const firstResult = wasm.decycler_js(close, hp_period, k);
    assert.strictEqual(firstResult.length, close.length);


    const secondResult = wasm.decycler_js(firstResult, hp_period, k);
    assert.strictEqual(secondResult.length, firstResult.length);


    const warmupEnd = 4;
    if (secondResult.length > warmupEnd + 10) {

        let hasValidValues = false;
        for (let i = warmupEnd; i < warmupEnd + 10; i++) {
            if (!isNaN(secondResult[i])) {
                hasValidValues = true;
                break;
            }
        }
        assert(hasValidValues, "Expected valid values after double application");
    }
});

test('Decycler all NaN input', () => {

    const allNaN = new Float64Array(100);
    allNaN.fill(NaN);

    assert.throws(() => {
        wasm.decycler_js(allNaN, 50, 0.707);
    }, /all values are nan/i);
});

test('Decycler batch single parameter set', () => {

    const close = new Float64Array(testData.close);


    const batchResult = wasm.decycler_batch(close, {
        hp_period_range: [125, 125, 0],
        k_range: [0.707, 0.707, 0]
    });


    const singleResult = wasm.decycler_js(close, 125, 0.707);

    assert.strictEqual(batchResult.values.length, singleResult.length);

    assertArrayClose(batchResult.values, singleResult, 5e-9, "Batch vs single mismatch");
});

test('Decycler batch multiple hp_periods', () => {

    const close = new Float64Array(testData.close.slice(0, 200));


    const batchResult = wasm.decycler_batch(close, {
        hp_period_range: [100, 150, 25],
        k_range: [0.707, 0.707, 0]
    });


    assert.strictEqual(batchResult.values.length, 3 * 200);
    assert.strictEqual(batchResult.rows, 3);
    assert.strictEqual(batchResult.cols, 200);


    const hp_periods = [100, 125, 150];
    for (let i = 0; i < hp_periods.length; i++) {
        const rowStart = i * 200;
        const rowEnd = rowStart + 200;
        const rowData = batchResult.values.slice(rowStart, rowEnd);

        const singleResult = wasm.decycler_js(close, hp_periods[i], 0.707);
        assertArrayClose(rowData, singleResult, 1e-10, `Row ${i} mismatch`);
    }
});

test('Decycler batch multiple parameters', () => {

    const close = new Float64Array(testData.close.slice(0, 50));


    const batchResult = wasm.decycler_batch(close, {
        hp_period_range: [10, 20, 10],
        k_range: [0.5, 0.7, 0.1]
    });


    assert.strictEqual(batchResult.values.length, 6 * 50);
    assert.strictEqual(batchResult.rows, 6);
    assert.strictEqual(batchResult.cols, 50);
    assert.strictEqual(batchResult.combos.length, 6);


    const expectedCombos = [
        { hp_period: 10, k: 0.5 },
        { hp_period: 10, k: 0.6 },
        { hp_period: 10, k: 0.7 },
        { hp_period: 20, k: 0.5 },
        { hp_period: 20, k: 0.6 },
        { hp_period: 20, k: 0.7 }
    ];

    for (let i = 0; i < expectedCombos.length; i++) {
        assert.strictEqual(batchResult.combos[i].hp_period, expectedCombos[i].hp_period,
            `hp_period mismatch at index ${i}`);
        assertClose(batchResult.combos[i].k, expectedCombos[i].k, 0.01,
            `k value mismatch at index ${i}`);
    }



    const rowIdx = 3;
    const rowStart = rowIdx * 50;
    const rowEnd = rowStart + 50;
    const rowData = batchResult.values.slice(rowStart, rowEnd);

    const singleResult = wasm.decycler_js(close, 20, 0.5);
    assertArrayClose(rowData, singleResult, 1e-10,
        `Batch row ${rowIdx} doesn't match single calculation`);
});

test('Decycler batch invalid parameters', () => {

    const close = new Float64Array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10]);


    assert.throws(() => {
        wasm.decycler_batch(close, {
            hp_period_range: [20, 10, 5],
            k_range: [0.707, 0.707, 0]
        });
    }, /invalid.*period|empty.*grid|invalid.*range|not enough valid data/i);




    assert.throws(() => {
        wasm.decycler_js(close, 5, -0.5);
    }, /invalid k/i);
});

test('Decycler batch edge cases', () => {

    const close = new Float64Array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10]);


    const singleBatch = wasm.decycler_batch(close, {
        hp_period_range: [5, 5, 1],
        k_range: [0.707, 0.707, 0.1]
    });

    assert.strictEqual(singleBatch.values.length, 10);
    assert.strictEqual(singleBatch.combos.length, 1);


    const largeBatch = wasm.decycler_batch(close, {
        hp_period_range: [5, 7, 10],
        k_range: [0.707, 0.707, 0]
    });


    assert.strictEqual(largeBatch.values.length, 10);
    assert.strictEqual(largeBatch.combos.length, 1);


    assert.throws(() => {
        wasm.decycler_batch(new Float64Array([]), {
            hp_period_range: [125, 125, 0],
            k_range: [0.707, 0.707, 0]
        });
    }, /empty input data|empty data/i);
});


test('Decycler zero-copy API basic', () => {

    const data = new Float64Array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10]);
    const hp_period = 5;
    const k = 0.707;


    const ptr = wasm.decycler_alloc(data.length);
    assert(ptr !== 0, 'Failed to allocate memory');


    const memView = new Float64Array(
        wasm.__wasm.memory.buffer,
        ptr,
        data.length
    );


    memView.set(data);


    try {
        wasm.decycler_into(ptr, ptr, data.length, hp_period, k);


        const regularResult = wasm.decycler_js(data, hp_period, k);
        for (let i = 0; i < data.length; i++) {
            if (isNaN(regularResult[i]) && isNaN(memView[i])) {
                continue;
            }
            assert.strictEqual(memView[i], regularResult[i],
                             `Value mismatch at index ${i}: ${memView[i]} vs ${regularResult[i]}`);
        }
    } finally {
        wasm.decycler_free(ptr, data.length);
    }
});

test('Decycler zero-copy separate buffers', () => {

    const data = new Float64Array([10, 20, 30, 40, 50, 60, 70, 80, 90, 100]);

    const inPtr = wasm.decycler_alloc(data.length);
    const outPtr = wasm.decycler_alloc(data.length);

    try {
        const inView = new Float64Array(wasm.__wasm.memory.buffer, inPtr, data.length);
        const outView = new Float64Array(wasm.__wasm.memory.buffer, outPtr, data.length);


        inView.set(data);


        wasm.decycler_into(inPtr, outPtr, data.length, 3, 0.707);


        for (let i = 0; i < data.length; i++) {
            assert.strictEqual(inView[i], data[i], `Input modified at index ${i}`);
        }


        const regularResult = wasm.decycler_js(data, 3, 0.707);
        for (let i = 0; i < data.length; i++) {
            if (isNaN(regularResult[i]) && isNaN(outView[i])) {
                continue;
            }
            assert.strictEqual(outView[i], regularResult[i],
                             `Output mismatch at index ${i}`);
        }
    } finally {
        wasm.decycler_free(inPtr, data.length);
        wasm.decycler_free(outPtr, data.length);
    }
});

test('Decycler zero-copy error handling', () => {



    assert.throws(() => {
        wasm.decycler_into(0, 0, 10, 125, 0.707);
    }, /null pointer|invalid memory/i);


    const ptr = wasm.decycler_alloc(10);
    try {

        assert.throws(() => {
            wasm.decycler_into(ptr, ptr, 10, 0, 0.707);
        }, /invalid.*period/i);


        assert.throws(() => {
            wasm.decycler_into(ptr, ptr, 10, 5, 0.0);
        }, /invalid k/i);
    } finally {
        wasm.decycler_free(ptr, 10);
    }
});


test('Decycler batch_into API', () => {

    const data = new Float64Array(testData.close.slice(0, 100));


    const expectedSize = 2 * 2 * 100;

    const inPtr = wasm.decycler_alloc(data.length);
    const outPtr = wasm.decycler_alloc(expectedSize);

    try {
        const inView = new Float64Array(wasm.__wasm.memory.buffer, inPtr, data.length);
        inView.set(data);


        const rows = wasm.decycler_batch_into(
            inPtr, outPtr, data.length,
            10, 20, 10,
            0.5, 0.7, 0.2
        );

        assert.strictEqual(rows, 4, 'Should have 4 parameter combinations');


        const regularBatch = wasm.decycler_batch(data, {
            hp_period_range: [10, 20, 10],
            k_range: [0.5, 0.7, 0.2]
        });

        const outView = new Float64Array(wasm.__wasm.memory.buffer, outPtr, expectedSize);
        assertArrayClose(
            Array.from(outView),
            regularBatch.values,
            1e-10,
            "Fast batch mismatch"
        );
    } finally {
        wasm.decycler_free(inPtr, data.length);
        wasm.decycler_free(outPtr, expectedSize);
    }
});


test('Decycler zero-copy memory management', () => {

    const sizes = [100, 1000, 10000, 100000];

    for (const size of sizes) {
        const ptr = wasm.decycler_alloc(size);
        assert(ptr !== 0, `Failed to allocate ${size} elements`);


        const memView = new Float64Array(wasm.__wasm.memory.buffer, ptr, size);
        for (let i = 0; i < Math.min(10, size); i++) {
            memView[i] = i * 1.5;
        }


        for (let i = 0; i < Math.min(10, size); i++) {
            assert.strictEqual(memView[i], i * 1.5, `Memory corruption at index ${i}`);
        }


        wasm.decycler_free(ptr, size);
    }
});


test('Decycler SIMD kernel consistency', () => {

    const data = new Float64Array(testData.close.slice(0, 100));
    const hp_period = 20;
    const k = 0.707;


    const autoResult = wasm.decycler_js(data, hp_period, k);



    assert.strictEqual(autoResult.length, data.length);


    const firstNonNaN = autoResult.findIndex(v => !isNaN(v));
    assert(firstNonNaN >= 2, 'Warmup period should have at least 2 NaN values');


    for (let i = firstNonNaN; i < autoResult.length; i++) {
        assert(!isNaN(autoResult[i]), `Unexpected NaN at index ${i} after warmup`);
    }
});

test.after(() => {
    console.log('Decycler WASM tests completed');
});

File: tests/python/test_decycler.py
"""
Python binding tests for Decycler indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
import sys
from pathlib import Path

try:
    import my_project as ta_indicators
except ImportError:
    
    try:
        import my_project as ta_indicators
    except ImportError:
        pytest.skip("Python module not built. Run 'maturin develop --features python' first", allow_module_level=True)

from test_utils import load_test_data, assert_close, EXPECTED_OUTPUTS
from rust_comparison import compare_with_rust


class TestDecycler:
    @pytest.fixture(scope='class')
    def test_data(self):
        return load_test_data()
    
    def test_decycler_partial_params(self, test_data):
        """Test Decycler with partial parameters (None values) - mirrors check_decycler_partial_params"""
        close = test_data['close']
        
        
        result = ta_indicators.decycler(close)  
        assert len(result) == len(close)
        
        
        result = ta_indicators.decycler(close, hp_period=50)
        assert len(result) == len(close)
        
        result = ta_indicators.decycler(close, hp_period=30, k=None)
        assert len(result) == len(close)
    
    def test_decycler_accuracy(self, test_data):
        """Test Decycler matches expected values from Rust tests - mirrors check_decycler_accuracy"""
        close = test_data['close']
        
        result = ta_indicators.decycler(
            close,
            hp_period=125,
            k=None  
        )
        
        assert len(result) == len(close)
        
        
        expected_last_5 = [
            60289.96384058519,
            60204.010366691065,
            60114.255563805666,
            60028.535266555904,
            59934.26876964316
        ]
        
        
        assert_close(
            result[-5:], 
            expected_last_5,
            rtol=1e-6,
            msg="Decycler last 5 values mismatch"
        )
    
    def test_decycler_default_params(self, test_data):
        """Test Decycler with default parameters"""
        close = test_data['close']
        
        
        result = ta_indicators.decycler(close)
        assert len(result) == len(close)
    
    def test_decycler_zero_period(self):
        """Test Decycler fails with zero period - mirrors check_decycler_zero_period"""
        input_data = np.array([10.0, 20.0, 30.0])
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.decycler(input_data, hp_period=0)
    
    def test_decycler_period_exceeds_length(self):
        """Test Decycler fails when period exceeds data length - mirrors check_decycler_period_exceed_length"""
        data_small = np.array([10.0, 20.0, 30.0])
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.decycler(data_small, hp_period=10)
    
    def test_decycler_very_small_dataset(self):
        """Test Decycler fails with insufficient data - mirrors check_decycler_very_small_dataset"""
        single_point = np.array([42.0])
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.decycler(single_point, hp_period=2)
    
    def test_decycler_empty_input(self):
        """Test Decycler fails with empty input"""
        empty = np.array([])
        
        with pytest.raises(ValueError, match="Empty data"):
            ta_indicators.decycler(empty)
    
    def test_decycler_invalid_k(self):
        """Test Decycler fails with invalid k"""
        data = np.array([1.0, 2.0, 3.0, 4.0, 5.0])
        
        with pytest.raises(ValueError, match="Invalid k"):
            ta_indicators.decycler(data, hp_period=2, k=0.0)
        
        with pytest.raises(ValueError, match="Invalid k"):
            ta_indicators.decycler(data, hp_period=2, k=-1.0)
        
        with pytest.raises(ValueError, match="Invalid k"):
            ta_indicators.decycler(data, hp_period=2, k=float('nan'))
    
    def test_decycler_reinput(self, test_data):
        """Test Decycler applied twice (re-input) - mirrors check_decycler_reinput"""
        close = test_data['close']
        
        
        first_result = ta_indicators.decycler(close, hp_period=30)
        assert len(first_result) == len(close)
        
        
        second_result = ta_indicators.decycler(first_result, hp_period=30)
        assert len(second_result) == len(first_result)
    
    def test_decycler_nan_handling(self, test_data):
        """Test Decycler handles NaN values correctly - mirrors check_decycler_nan_handling"""
        close = test_data['close']
        
        result = ta_indicators.decycler(close, hp_period=125)
        assert len(result) == len(close)
        
        
        if len(result) > 240:
            assert not np.any(np.isnan(result[240:])), "Found unexpected NaN after warmup period"
    
    def test_decycler_warmup_period(self, test_data):
        """Test Decycler warmup period matches expected behavior"""
        close = test_data['close']
        hp_period = 125
        
        result = ta_indicators.decycler(close, hp_period=hp_period)
        
        
        first_non_nan = np.where(~np.isnan(result))[0]
        if len(first_non_nan) > 0:
            first_idx = first_non_nan[0]
            
            
            first_input = np.where(~np.isnan(close))[0][0]
            expected_warmup = first_input + 2
            
            assert first_idx == expected_warmup, f"Warmup period mismatch: expected {expected_warmup}, got {first_idx}"
            
            
            assert np.all(np.isnan(result[:expected_warmup])), "Expected NaN values during warmup period"
    
    def test_decycler_partial_nan_input(self, test_data):
        """Test Decycler with data containing some NaN values"""
        
        data = np.arange(1.0, 101.0)  
        
        
        data_with_nan = data.copy()
        data_with_nan[10] = np.nan
        data_with_nan[11] = np.nan
        
        
        result = ta_indicators.decycler(data_with_nan, hp_period=5)
        assert len(result) == len(data_with_nan)
        
        
        
        
    
    def test_decycler_edge_case_k_values(self):
        """Test Decycler with edge case k values"""
        data = np.arange(1.0, 101.0)  
        
        
        result_small = ta_indicators.decycler(data, hp_period=10, k=0.001)
        assert len(result_small) == len(data)
        
        
        result_default = ta_indicators.decycler(data, hp_period=10, k=0.707)
        assert len(result_default) == len(data)
        
        
        result_large = ta_indicators.decycler(data, hp_period=10, k=10.0)
        assert len(result_large) == len(data)
        
        
        
        warmup_end = 12  
        if len(data) > warmup_end + 10:
            check_idx = warmup_end + 5
            assert result_small[check_idx] != result_default[check_idx], "Different k values should produce different results"
            assert result_default[check_idx] != result_large[check_idx], "Different k values should produce different results"
    
    def test_decycler_input_types(self, test_data):
        """Test Decycler with different input types"""
        close = test_data['close'][:100]  
        
        
        close_list = list(close)
        result_list = ta_indicators.decycler(np.array(close_list), hp_period=30)
        assert len(result_list) == len(close)
        
        
        close_f64 = np.array(close, dtype=np.float64)
        result_f64 = ta_indicators.decycler(close_f64, hp_period=30)
        assert len(result_f64) == len(close)
        
        
        close_f32 = np.array(close, dtype=np.float32)
        result_f32 = ta_indicators.decycler(close_f32.astype(np.float64), hp_period=30)
        assert len(result_f32) == len(close)
        
        
        assert_close(result_f32, result_f64, rtol=1e-6, msg="Results should be very similar after conversion")
    
    def test_decycler_streaming(self, test_data):
        """Test Decycler streaming matches batch calculation"""
        close = test_data['close']
        hp_period = 125
        k = 0.707
        
        
        batch_result = ta_indicators.decycler(close, hp_period=hp_period, k=k)
        
        
        stream = ta_indicators.DecyclerStream(hp_period=hp_period, k=k)
        stream_values = []
        
        for price in close:
            result = stream.update(price)
            stream_values.append(result if result is not None else np.nan)
        
        stream_values = np.array(stream_values)
        
        
        assert len(batch_result) == len(stream_values)
        
        
        
        first_input = np.where(~np.isnan(close))[0][0]
        expected_warmup = first_input + 2
        
        
        for i in range(expected_warmup):
            assert np.isnan(batch_result[i]), f"Expected NaN in batch during warmup at index {i}"
            assert np.isnan(stream_values[i]), f"Expected NaN in stream during warmup at index {i}"
        
        
        for i in range(expected_warmup, len(batch_result)):
            if np.isnan(batch_result[i]) and np.isnan(stream_values[i]):
                continue
            if np.isnan(batch_result[i]) or np.isnan(stream_values[i]):
                
                continue
            assert_close(batch_result[i], stream_values[i], rtol=1e-9, atol=1e-9, 
                        msg=f"Decycler streaming mismatch at index {i}")
    
    def test_decycler_batch(self, test_data):
        """Test Decycler batch processing - mirrors check_batch_default_row"""
        close = test_data['close']
        
        result = ta_indicators.decycler_batch(
            close,
            hp_period_range=(125, 125, 0),  
            k_range=(0.707, 0.707, 0.0)  
        )
        
        assert 'values' in result
        assert 'hp_periods' in result
        assert 'ks' in result
        assert 'rows' in result
        assert 'cols' in result
        
        
        assert result['values'].shape[0] == 1
        assert result['values'].shape[1] == len(close)
        
        
        default_row = result['values'][0]
        expected_last_5 = [
            60289.96384058519,
            60204.010366691065,
            60114.255563805666,
            60028.535266555904,
            59934.26876964316
        ]
        
        
        assert_close(
            default_row[-5:],
            expected_last_5,
            rtol=1e-6,
            msg="Decycler batch default row mismatch"
        )
    
    def test_decycler_batch_multiple_params(self, test_data):
        """Test Decycler batch with multiple parameter combinations"""
        close = test_data['close']
        
        result = ta_indicators.decycler_batch(
            close,
            hp_period_range=(100, 150, 25),  
            k_range=(0.5, 0.7, 0.1)  
        )
        
        
        assert result['values'].shape[0] == 9
        assert result['values'].shape[1] == len(close)
        assert len(result['hp_periods']) == 9
        assert len(result['ks']) == 9
        
        
        expected_hp_periods = [100, 100, 100, 125, 125, 125, 150, 150, 150]
        expected_ks = [0.5, 0.6, 0.7, 0.5, 0.6, 0.7, 0.5, 0.6, 0.7]
        
        for i, (hp, k) in enumerate(zip(expected_hp_periods, expected_ks)):
            assert result['hp_periods'][i] == hp, f"hp_period mismatch at index {i}"
            assert_close(result['ks'][i], k, rtol=1e-9, msg=f"k value mismatch at index {i}")
        
        
        
        row_idx = 4
        single_result = ta_indicators.decycler(close, hp_period=125, k=0.6)
        assert_close(
            result['values'][row_idx], 
            single_result, 
            rtol=1e-9,
            msg=f"Batch row {row_idx} doesn't match single calculation"
        )
    
    def test_decycler_all_nan_input(self):
        """Test Decycler with all NaN values"""
        all_nan = np.full(100, np.nan)
        
        with pytest.raises(ValueError, match="decycler: All values are NaN"):
            ta_indicators.decycler(all_nan, hp_period=50)  
    
    def test_decycler_kernel_parameter(self, test_data):
        """Test Decycler with different kernel parameters"""
        close = test_data['close']
        
        
        result_scalar = ta_indicators.decycler(close, kernel="scalar")
        assert len(result_scalar) == len(close)
        
        
        result_auto = ta_indicators.decycler(close, kernel=None)
        assert len(result_auto) == len(close)
        
        
        with pytest.raises(ValueError, match="Unknown kernel"):
            ta_indicators.decycler(close, kernel="invalid")


if __name__ == '__main__':
    pytest.main([__file__, '-v'])